sip/sip-004-materialized-view.md

```diff
@@ -0,0 +1,822 @@
+a peer will visit when resolving a back-pointer can be be at most _O(log B)_ -- i.e.
+this would only happen if the blockchain's forks were organized into a
+perfectly-balanced binary tree. 
+
+In practice, there will be one _long_ fork row that encodes the canonical
+history, as well as number of short fork rows that encode short-lived forks
+(which can arise naturally from burn chain reorganizations).  This means
+resolving back-pointers while working on the longest fork -- the fork where a
+miner's block rewards are most likely to be realized -- will be _O(1)_ in
+expectation.  To help achieve this, the ancestor table would be implemented as
+a hash table in order to ensure that finding the ancestor block also runs in
+_O(1)_ time.
+
+### Merklized Skip-list
+
+The third principal data structure in a MARF is a Merklized skip-list encoded
+from the block header hashes and ART root hashes in each block.  The hash of the
+root node in the ART for block _N_ is derived not only from the hash of the
+root's children, but also from the hashes of the block headers from blocks
+`N - 1`, `N - 2`, `N - 4`, `N - 8`, `N - 16`, and so on.  This constitutes
+a _Merklized skip-list_ over the sequence of ARTs.
+
+The reason for encoding the root node's hash this way is to make it possible for
+peers to create a cryptographic proof that a particular key maps to a particular
+value when the value lives in a prior block, and can only be accessed by
+following one or more back-pointers.  In addition, the Merkle skip-list affords
+a client _two_ ways to verify key-value pairs:  the client only needs either (1)
+a known-good root hash, or (2) the sequence of block headers for the Stacks
+chain and its underlying burn chain.  Having (2) allows the client to determine
+(1), but calculating (2) is expensive for a client doing a small number of
+queries.  For this reason, both options are supported.
+
+### MARF Merkle Proofs
+
+A Merkle proof for a MARF is constructed using a combination of two types of
+sub-proofs:  _segment proofs_, and _shunt proofs_.  A _segment proof_ is a proof
+that a node belongs to a particular Merklized ART.  It is simply a Merkle tree
+proof.  A _shunt proof_ is a proof that the ART for block _N_ is exactly _K_
+blocks away from the ART at block _N - K_.  It is generated as a Merkle proof
+from the Merkle skip-list.
+
+Calculating a MARF Merkle proof is done by first calculating a segment proof for a
+sequence of path prefixes, such that all the nodes in a single prefix are in the
+same ART.  To do so, the node walks from the current block's ART's root node
+down to the leaf in question, and each time it encounters a back-pointer, it
+generates a segment proof from the _currently-visited_ ART to the intermediate
+node whose child is the back-pointer to follow.  If a path contains _i_
+back-pointers, then there will be _i+1_ segment proofs.
+
+Once the peer has calculated each segment proof, it calculates a shunt proof
```

Comment:
If I understand this correctly, the goal of the shunt proofs is to prove:

1. For all blocks _b_i_ in the segment proofs, _b_i+1_ is the block _k_ steps prior to _b_i_, where _k_ is whatever the back-pointer in the segment proof of _b_i_ indicated.
2. For all block _b_i_ in the segment proofs, _b_i_ is included in the same fork as _b_0_ where _b_0_ is the chain tip.

If that's true, is proving (1) strictly necessary? The fact that the MARF commits these back-pointers to blockhashes means that the referenced block is already committed to, regardless of how many steps back it was, right?

In either case, this is just a normal Merkle skip-list proof, right, with the difference being that rather than _just_ proving

_b_0_ -> _b_genesis_

We want to generate a proof that

_b_0_ -> _b_1_ -> _b_2_ -> ... -> _b_genesis

And the skip-listiness of the proof should _already_ indicate how many blocks separate each entry, right? I may be misremembering some of the properties that skip-lists give you.

Reply:
Either way, the verifier needs to know the hashes of blocks _b_i_ to verify the proof.  However, the verifier must also verify that:

1. these blocks exist
2. these blocks are in the same order on the burn chain
3. these blocks belong to the same fork

The verifier _could_ verify a proof without the prover having to construct shunt proofs by building and maintaining its own fork table.  To do so, the verifier would need to _a priori_ fetch the burn chain and the Stacks chain headers.  Then, all the prover would need to supply are indexes _i_ to the blocks to which the intermediate nodes' backpointers were followed, and the verifier could fill in the rest.

Unfortunately, having verifiers fetch both chain headers is a tall order for low-power verifiers, like those within a Web browser or on a mobile phone.  The advantage of proving both (1) and (2) in your question is that it removes the need for the verifier to build its own fork table.  Instead, it can use a commodity burn chain block explorer to spot-check that Stacks block headers _b_i_ correspond to sortition transactions in the burn chain, and are the given number of blocks apart.

Reply:
```
these blocks exist
these blocks are in the same order on the burn chain
these blocks belong to the same fork
```

Wouldn't all of those properties be proven in a simple Merkle skip-list proof that includes the blocks in question (and the genesis)? It proves that the blocks are all in the same fork, because otherwise, there wouldn't be a skip-list path. Same with ordering and existence.

Reply:
I guess I'm confused by what you're asking?  You were questioning why do we prove the following:

```
For all blocks b_i in the segment proofs, b_i+1 is the block k steps prior to b_i, where k is whatever the back-pointer in the segment proof of b_i indicated.
```

Isn't this exactly a Merkle skip-list proof that includes the blocks in question?  In order to include an arbitrary sequence of blocks, you would need several Merkle skip-list proofs -- one that connects _b_i_ to _b_i+1_ for each _i_, since it is usually not the case that _b_i+1_ will have a direct connection to _b_i_.

Reply:
Right, what I'm trying to get at is understanding why we need to perform _n_ different shunts proofs, instead of a single proof that visits the _n_ required nodes. For example, if we visit a single back-pointer in the construction of the proof, that means, we have two blocks and the genesis that absolutely must be visited:

B_0 = block 256, has skip list pointers to n = 255, 254, 252, 248, 240, 224, 192, 128, 0
B_1 = block 197, has skip list pointers to n = 196, 195, 193, 189, 181, 165, 133, 69, 0

So the proof would be composed of headers for:
256 -> 224 -> 208 -> 200 -> 198 -> 197 -> 0

Which seems to be much less than what is described in the SIP -- I'm probably missing something.

Reply:
```
B_0 = block 256, has skip list pointers to n = 255, 254, 252, 248, 240, 224, 192, 128, 0
B_1 = block 197, has skip list pointers to n = 196, 195, 193, 189, 181, 165, 133, 69, 0

So the proof would be composed of headers for:
256 -> 224 -> 208 -> 200 -> 198 -> 197 -> 0
```

B_1 would not include a skip-list pointer for 0, since the hash of block 0 wouldn't have gone into the hash for block 197.

But regardless, you're trying to prove that block 256 is a descendant of block 197.  In order to do this, you'd need to (1) walk back from 256 to 197 along the skip list, and (2) for each node visited on the walk, supply the sequence of hashes that went into producing its hash.

To prove that 256 descended from 197, you'd need the following sequence of hashes:

* [255, 254, 252, 248, 240, 224, 192, 128, 0]  // walk from 256 to 224
* [224, 223, 222, 220, 216, 208, 192, 160, 96]  // walk from 224 to 208
* [208, 207, 206, 204, 200, 192, 176, 144, 80]  // walk from 208 to 200
* [200, 199, 198, 196, 192, 184, 168, 136, 72]  // walk from 200 to 198
* [198, 197, 196, 194, 190, 182, 166, 134, 70]  // you're done -- you've landed on 197

In practice, you'd omit some of the above hashes, since they can be calculated on-the-fly (and you'd want to verify that the _i_th hash was present in the right place for the hashes that go into calculating the _i+1_th hash).

It's a little trickier when you supply segment proofs, since the node's hash is calculated from both the segment proof and the last connecting shunt proof (the code calls it a "junction" proof).  But my point is, if you want to supply a proof that _A_ is reachable from _B_ in a Merkle skip-list, you need to supply the hashes that went into calculating the roots of all nodes visited on the walk.

Reply:
I see -- I think I was just confused by this description. Yes, all of those hashes are included in the proof, but really that's just because they are the composite values of the skip list node (so if you want to verify the hash provided for that node, you need to have those "pruned" hashes). The only nodes being "visited" in the proof (that is, the only path being traversed), is the one from the chain tip to the genesis block, which must include every block that was in the segmented proofs, correct ?

Reply:
> The only nodes being "visited" in the proof (that is, the only path being traversed), is the one from the chain tip to the genesis block, which must include every block that was in the segmented proofs, correct ?

Yes.  But, the requirement that it "must include every block that was in the segmented proof" means that you will need to supply a sequence of sequences of hashes between two successive block hashes.  This is because you will almost certainly visit one or more intermediate blocks between blocks _i_ and _i+1_ while walking the skip-list, and you need to show that _those_ blocks exist as part of the fork as well.  That is, if you have the root hash for block 256 already, but don't have the root hash for 197, then the reason I can't just supply you the hashes for blocks 256, 224, 208, 200, 198, 197, and 0, but must instead supply all of the additional hashes I mentioned earlier, is because this isn't enough to convince you that block 208 was an ancestor of block 256.  You _do_ however know that 224 was an ancestor, because you know the hash of 256 and I can supply you the sequence of hashes that went into calculating it.  I can convince you that 224 is an ancestor of 256 this way, since I can't brute-force a hash-collision.  You can then ask me for the hashes that went into producing 224's hash, and from there, continue you walk back to 197.  This is why the Merkle proof for a MARF needs shunt proofs -- it's the standard way to prove that _A_ is the ancestor of _B_ in a Merkle skip-list for any nodes _A_ < _B_.

Reply:
Yes, but that's true for all Merkle proofs of any kind: you need to be able to recompute a node's hash using some pruned children (in this case, that's the block hashes that lead from 224 back to genesis) and some data that you are proving (in this case, the block hash of 208). Looking at the segment proofs for example, you don't describe the pruned sibling hashes this way, even though those too need to be included in the proof.

Reply:
That is correct.  The hash of each node is the hash of its children, plus the hash of the geometric series of ancestor hashes.  Thus to verify a node, you need the hashes of its ancestors and the hash of its children.  You calculate the former from the shunt proof, and the latter from the segment proof.

I didn't go into detail on how to calculate a segment proof because I thought the audience would understand how a Merkle tree proof worked already.  The novel aspect of MARF Merkle proofs is that they incorporate shunt proofs between segment proofs, and have a different way of calculating the trie root hash.  Happy to go back and add a more detailed description if you feel that it would be appropriate!

Reply:
> Happy to go back and add a more detailed description if you feel that it would be appropriate!

I think actually the opposite! I thought the segment proofs were described perfectly, but ended up getting lost in the shunts, and think it would probably be more clear with less detail. Ultimately, if someone is interested in a full enumeration of what goes into the proof construction, they should look at the code -- this SIP should give them enough of an idea about what's different with the shunt proof (which is the fact that it must prove a link between all of the nodes in the segment proofs). Ultimately, anyone writing a verifier or separate implementation would end up needing to figure out what goes into the hash functions anyways.